### ISSUE TITLE
composite codebooks and (round, channel) limitations

### ISSUE THREAD

**Issue by berl** [spacetx/starfish#1121]
Barcoded multiplexing methods increase the number of mRNA species read out per round by using non-diagonal codebooks. This multiplexing combines with spatial resolution limitations of light microscopy to place an upper limit on the expression levels (spots per cell) that can be measured.  To increase the range of expression levels measured in a given experiment, published results in the field (MERFISH, SeqFISH) have included what I'm calling "composite codebooks" with both barcoded and diagonal blocks. The diagonal blocks are read out with a single gene per `(round, channel)`, reducing the impact of spatial resolution on the accessible expression levels.

I see two challenges in `starfish` to processing composite codebooks in a single `starfish` `Experiment`, and I'll formulate them as questions in hopes of starting a conversation

1. If a method uses different spotfinding or decoding methods to handle the diagonal blocks, can `starfish` pipeline processing be parameterized to handle different parts of a codebook with different processing methods?

2. For various reasons (humor me here) the diagonal block of a composite codebook could be read out with a different number of channels than the barcoded block, can `starfish` handle a situation where there are different numbers of channels in different rounds?  

Workarounds/comments:  
I think 1. can be addressed with two different `Experiment`s per experiment. 2. can also be addressed in the same way as long as there are the same number of channels per round in each block.
 
these are both suboptimal because they break a scientific experiment into multiple  `Experiment`s and require additional work to track registration and metadata. 2.) really just highlights a limitation of our original spacetx spec. Once we defined numbers of rounds and channels as fixed quantities for the `Experiment`, we're sort of stuck.  A more general workaround would map the acquisition `(r,c)` to a single `round` with N channels, where `N != n_rounds * n_channels`

**Comment by berl** [comment]
cc: @njmei

**Comment by njmei** [comment]
Some toy example failure cases:

1. Uneven number of channels for primary images
```
Round 0:
    Channel 0: primary_probe_0
    Channel 1: primary_probe_1
    Channel 2: primary_probe_2
    Channel 3: primary_probe_3
Round 1:
    Channel 0: primary_probe_4
    Channel 1: primary_probe_5
    Channel 2: dapi_aux_probe
    Channel 3: beads_aux_probe
```

2. Uneven number of channels for primary images per round and non-contiguous aux images (of the same type)
```
Round 0:
    Channel 0: primary_probe_0
    Channel 1: primary_probe_1
    Channel 2: primary_probe_2
    Channel 3: dapi_aux_probe
Round 1:
    Channel 0: primary_probe_3
    Channel 1: primary_probe_4
    Channel 2: primary_probe_5
    Channel 3: primary_probe_6
Round 2:
    Channel 0: primary_probe_7
    Channel 1: primary_probe_8
    Channel 2: primary_probe_9
    Channel 3: dapi_aux_probe
```

2. Unlikely, but food for thought
```
Round 0:
    Channel 0: primary_probe_0
    Channel 1: primary_probe_1
    Channel 2: actin_aux_probe
    Channel 3: dapi_aux_probe
Round 1:
    Channel 0: primary_probe_2
    Channel 1: actin_aux_probe
    Channel 2: primary_probe_3
    Channel 3: primary_probe_4
Round 2:
    Channel 0: dapi_aux_probe
    Channel 1: primary_probe_5
    Channel 2: primary_probe_6
    Channel 3: primary_probe_7
```

**Comment by berl** [comment]
@njmei thanks for the concrete examples of the problems in Question 2.  Also thanks (?) for the added wrinkle of `aux` channels!

 @dganguli and @ambrosejcarr thoughts? 
right now @njmei's examples are pretty effective examples of how not to use `starfish` (indeed, how not to acquire data if you want to use `starfish`).  They highlight a lack of generality in our original model that may be worth exploring further.

**Comment by ambrosejcarr** [comment]
Interesting issue. Can split codebook images be processed with the same image analysis pipeline (including the same parameters)? 

If yes, my gut feeling is that these types of experiments should have two codebooks, and two decoding steps should be run sequentially. The resulting IntensityTables can be concatenated.  

If the same parameterization does not work for both acquisition strategies, my sense is that you'd need to run two different analysis pipelines (or the pipeline would need to branch) and in this case I think splitting the codebook still makes sense. 

Regarding the failure cases, can you say more? Have you tried these? How does starfish break?

**Comment by njmei** [comment]
@ambrosejcarr 

Hi Ambrose, I haven't literally tried the example test cases but know they fail pretty much right from the get-go because starfish treats `rounds` and `channels` as independent dimensions. There might be further downstream issues with the examples.

1. When trying to format the incoming example:
```
primary_image_dimensions: Mapping[Union[str, Axes], int] = 
{
        Axes.ROUND: 2,
        Axes.CH: 4, #starfish will fail when trying to load (Round 1, Channel 2) or (Round 1, Channel 3)
        Axes.ZPLANE: 50,
}
```

2. Again, when trying to format the incoming example:
```
primary_image_dimensions =
{
    Axes.ROUND: 3,
    Axes.CH: 4, #starfish will do the wrong thing by trying to load (Round 0, Channel 3) and (Round 2, Channel 4)
    Axes.ZPLANE: 50
}

dapi_aux_image_dimensions =
{
    Axes.ROUND: Unclear what this should be since the DAPI rounds are in Round 0 and Round 2... (If a value of '2' is used, then starfish loses information about how aux rounds relate to primary rounds, which may be useful for applying alignments between aux rounds to primary images)
    Axes.CH: Again, the number of channels can be either 1, or 0 depending on which round of the experiment we're in.
    Axes.ZPLANE: 50
}
```

3. This example is one where even the channel identity of a 'type' of aux image is not consistent (on top of the above issues). We have a mapping strategy that can partially deal with the issue, but it's specific to our `manifest` metadata format so if other users ever run a similar experiment, they will be out of luck with Starfish.

**Comment by ambrosejcarr** [comment]
@simone-codeluppi confirmed that this is an issue for his datasets as well, and confirmed that he is unable to construct and load experiment objects given our existing restrictions.

**Comment by dganguli** [comment]
These are all very good points and definitely deserve some careful thought/discussion. I think there's a few things going on here that I'll try to re-articulate to make sure I understand:

1. Assuming **primary** images fit into a R x C x Z x Y x X tensor is a poor assumption for **non-multiplexed assays**. It is, however, a good assumption for multiplexed assays. Although see point 3 below.

This can probably be mitigated by storing the image planes/volumes in a dictionary like data structure instead of an xarray. Or, we can put zeros in the xarray where a certain R,C combo doesn't have real data. The latter idea smells a little gross to me though.

2. Assuming **auxiliary** images fit in a R x C x Z x Y x X tensor is a poor assumption for both multiplexed and non-multiplexed assays. 

This can also be mitigated by storing image planes/volumes in a dictionary like structure instead of an xarray.

3. Multiplexed assays (e.g., SEQFISH and MERFISH) often have a non-multiplexed data acquisition stage, e.g., to deal with highly expressed transcripts, as a control, etc. When I've looked at the MERFISH and SEQFISH code, the code branches at this point according to some hardcoded (and typically inscrutable) logic. The image processing is typically the same, but the decoding is obviously different. 

I think that having two Experiments and two codebooks, united under a top level Experiment (i *think this is possible), is not a bad idea for how to process these data. The pipeline recipe will also be two fold, with some shared steps, and some different steps. In the end, the resulting 2 intensity tables will be concatenated.

**Comment by dganguli** [comment]
Hi all, @ttung and I just re-visited this issue. Points 1 and 2 in my above re-articulation of the points raised in this issue, are now addressed by the proposal in #1322. Please chime in there and let us know what you think @njmei ! 

For Point 3, @ttung and I also believe the simplest and cleanest solution  is to have two separate experiments, with two separate recipes and codebooks. It's up to the user to coordinate metadata and results across these experiments. This has also been suggested by @berl and @ambrosejcarr above.

As such, I vote we close this issue. @shanaxel42 @neuromusic FYI

**Comment by neuromusic** [comment]
Can we get some documentation clarifying (or better: demonstrating) the
strategy for 3?

**Comment by shanaxel42** [comment]
closing, will be addressed with #1368
